File: python/src/circle_ai/core/circle_engine.py
# core/circle_engine.py
#
# Port of:
#   • CircleAI.Core.CircleEngine        — top-level facade + type-keyed module bag
#   • CircleAI.Core.ICircleModule       — attachable module contract
#   • CircleAI.Core.IEmbeddingService   — embedding module contract
#
# The C# module bag is keyed on the CLR generic type argument (typeof(T)).
# Python has no reified generics, so register_module / get_module / has_module
# take the key type explicitly. When omitted on register, the concrete type of
# the instance is used — the common case.

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Dict, Optional, Type, TypeVar

from .model_loader import IModelLoader

T = TypeVar("T")


# ─────────────────────────────────────────────────────────────────────────────
# ICircleModule — attachable module contract.
# ─────────────────────────────────────────────────────────────────────────────


class ICircleModule(ABC):
    """A module that attaches to a :class:`CircleEngine`.

    Disposable — hosts call :meth:`dispose` on teardown.
    """

    @property
    @abstractmethod
    def module_name(self) -> str:
        """Human-readable module name."""
        raise NotImplementedError

    @abstractmethod
    async def init_async(self, engine: "CircleEngine") -> None:
        """Initialise the module against the owning engine."""
        raise NotImplementedError

    @property
    @abstractmethod
    def is_model_loaded(self) -> bool:
        """True once the module's backing model is loaded."""
        raise NotImplementedError

    def dispose(self) -> None:
        """Release resources. Default is a no-op; override as needed."""
        return None


# ─────────────────────────────────────────────────────────────────────────────
# IEmbeddingService — embedding module contract.
# ─────────────────────────────────────────────────────────────────────────────


class IEmbeddingService(ICircleModule, ABC):
    """An :class:`ICircleModule` that produces dense embeddings."""

    @abstractmethod
    def generate_embedding(self, text: str) -> list[float]:
        """Embed *text* into a dense float vector."""
        raise NotImplementedError

    @property
    @abstractmethod
    def embedding_size(self) -> int:
        """Number of floats each embedding carries."""
        raise NotImplementedError


# ─────────────────────────────────────────────────────────────────────────────
# CircleEngine — top-level facade.
# ─────────────────────────────────────────────────────────────────────────────


class CircleEngine:
    """Top-level facade for the CircleAI on-device stack.

    Holds the :class:`IModelLoader` and a small type-keyed registry of attached
    modules (embeddings, search, chat generators, tool bridges) wired in from
    downstream packages.
    """

    __slots__ = ("_model_loader", "_modules", "embedding_service")

    def __init__(self, model_loader: IModelLoader) -> None:
        if model_loader is None:
            raise ValueError("model_loader")
        self._model_loader = model_loader
        self._modules: Dict[type, object] = {}
        # Optional embedding service — wired in by the embeddings package.
        # Kept as a plain attribute so Core need not reference downstream impls.
        self.embedding_service: Optional[object] = None

    @property
    def model_loader(self) -> IModelLoader:
        """The model loader used to acquire and cache model files."""
        return self._model_loader

    def register_module(self, module: T, key_type: Optional[Type] = None) -> "CircleEngine":
        """Register a module instance keyed by ``key_type`` (defaults to the
        instance's concrete type). Returns self for chaining."""
        if module is None:
            raise ValueError("module")
        key = key_type if key_type is not None else type(module)
        self._modules[key] = module
        return self

    def get_module(self, key_type: Type[T]) -> Optional[T]:
        """Retrieve a previously registered module, or ``None`` if none was
        registered for that type."""
        return self._modules.get(key_type)  # type: ignore[return-value]

    def has_module(self, key_type: Type) -> bool:
        """True if a module of the given type has been registered."""
        return key_type in self._modules

File: swift/Sources/CircleAI/CircleEngine.swift
// CircleEngine.swift
//
// Port of CircleAI.Core.CircleEngine + ICircleModule + IEmbeddingService.
//
// Top-level facade for the CircleAI on-device stack. Holds the IModelLoader
// and a small registry of attached modules (embeddings, search, chat
// generators, tool bridges) wired in from downstream code via extensions.
//
// CircleAI.Core deliberately knows nothing about Inference / Embeddings /
// Search / Tools. Those attach their own services through the
// registerModule / getModule pair, or through settable properties such as
// `embeddingService`.

import Foundation

// MARK: - ICircleModule — CircleAI.Core.ICircleModule

/// A pluggable CircleAI module. `IDisposable` in C# maps to an explicit
/// `dispose()`; conformers own native/model state that must be released.
public protocol ICircleModule: AnyObject {
    var moduleName: String { get }
    func initAsync(engine: CircleEngine) async throws
    var isModelLoaded: Bool { get }
    func dispose()
}

// MARK: - IEmbeddingService — CircleAI.Core.IEmbeddingService

/// An embedding module. Extends `ICircleModule` (as in C#) and adds
/// synchronous embedding generation plus the vector size.
public protocol IEmbeddingService: ICircleModule {
    func generateEmbedding(_ text: String) -> [Float]
    var embeddingSize: Int { get }
}

// MARK: - CircleEngine — CircleAI.Core.CircleEngine

/// Top-level facade. A constructor that takes an `IModelLoader`, a public
/// `modelLoader`, and a type-keyed module bag.
///
/// Stateful (the module bag mutates), so a `final class ... @unchecked
/// Sendable` guarded by an `NSLock`. The lock is confined to the synchronous
/// mutation helpers.
public final class CircleEngine: @unchecked Sendable {
    private let lock = NSLock()
    private var modules: [ObjectIdentifier: Any] = [:]

    /// The model loader used to acquire and cache model files.
    public let modelLoader: any IModelLoader

    /// Optional embedding service. Wired in by an embeddings extension. Kept as
    /// `Any?` so Core does not need to reference downstream implementations.
    private var _embeddingService: Any?
    public var embeddingService: Any? {
        get { lock.lock(); defer { lock.unlock() }; return _embeddingService }
        set { lock.lock(); _embeddingService = newValue; lock.unlock() }
    }

    public init(modelLoader: any IModelLoader) {
        self.modelLoader = modelLoader
    }

    /// Register a module instance keyed by its concrete or protocol type `T`.
    @discardableResult
    public func registerModule<T>(_ module: T, as type: T.Type = T.self) -> CircleEngine {
        lock.lock()
        modules[ObjectIdentifier(type)] = module
        lock.unlock()
        return self
    }

    /// Retrieve a previously registered module, or `nil` if none was
    /// registered for that type.
    public func getModule<T>(_ type: T.Type = T.self) -> T? {
        lock.lock(); defer { lock.unlock() }
        return modules[ObjectIdentifier(type)] as? T
    }

    /// Returns true if a module of the given type has been registered.
    public func hasModule<T>(_ type: T.Type = T.self) -> Bool {
        lock.lock(); defer { lock.unlock() }
        return modules[ObjectIdentifier(type)] != nil
    }
}
